src/parcsr_ls/par_ilu.c

```diff
@@ -2296,173 +2423,187 @@ hypre_ILULocalRCM( hypre_CSRMatrix *A, HYPRE_Int start, HYPRE_Int end,
    rqperm = hypre_TAlloc(HYPRE_Int, n, HYPRE_MEMORY_HOST);
-   for (i = 0 ; i < n ; i ++)
+   for (i = 0; i < n; i++)
    {
       rqperm[qperm[i]] = i;
    }
+
+   /* Set/Move A_i and A_j to host */
+   if (hypre_GetActualMemLocation(memory_location) == hypre_MEMORY_DEVICE)
+   {
+      A_i = hypre_TAlloc(HYPRE_Int, n + 1, HYPRE_MEMORY_HOST);
+      A_j = hypre_TAlloc(HYPRE_Int, A_nnz, HYPRE_MEMORY_HOST);
+
+      hypre_TMemcpy(A_i, hypre_CSRMatrixI(A), HYPRE_Int, n + 1,
+                    HYPRE_MEMORY_HOST, HYPRE_MEMORY_DEVICE);
+      hypre_TMemcpy(A_j, hypre_CSRMatrixJ(A), HYPRE_Int, A_nnz,
+                    HYPRE_MEMORY_HOST, HYPRE_MEMORY_DEVICE);
+   }
+   else
+   {
+      A_i = hypre_CSRMatrixI(A);
+      A_j = hypre_CSRMatrixJ(A);
+   }
+
    /* 2: Build Graph
     * Build Graph for RCM ordering
     */
-   G = hypre_CSRMatrixCreate(num_nodes, num_nodes, 0);
-   hypre_CSRMatrixInitialize(G);
-   hypre_CSRMatrixSetDataOwner(G, 1);
-   G_i = hypre_CSRMatrixI(G);
-   if (sym)
-   {
-      /* Directly use A */
-      G_nnz = 0;
-      G_capacity = hypre_max(A_nnz * n * n / num_nodes / num_nodes - num_nodes, 1);
-      G_j = hypre_TAlloc(HYPRE_Int, G_capacity, HYPRE_MEMORY_DEVICE);
-      for (i = 0 ; i < num_nodes ; i ++)
-      {
-         G_i[i] = G_nnz;
-         row = perm[i + start];
-         r1 = A_i[row];
-         r2 = A_i[row + 1];
-         for (j = r1 ; j < r2 ; j ++)
+   G_nnz = 0;
+   G_capacity = hypre_max((A_nnz * n * n / num_nodes / num_nodes) - num_nodes, 1);
+   G_i = hypre_TAlloc(HYPRE_Int, num_nodes + 1, HYPRE_MEMORY_HOST);
+   G_j = hypre_TAlloc(HYPRE_Int, G_capacity, HYPRE_MEMORY_HOST);
+
+   /* TODO (VPM): Extend hypre_CSRMatrixPermute to replace the block below */
+   for (i = 0; i < num_nodes; i++)
+   {
+      G_i[i] = G_nnz;
+      row = perm[i + start];
+      r1 = A_i[row];
+      r2 = A_i[row + 1];
+      for (j = r1; j < r2; j ++)
+      {
+         col = rqperm[A_j[j]];
+         if (col != row && col >= start && col < end)
          {
-            col = rqperm[A_j[j]];
-            if (col != row && col >= start && col < end)
+            /* this is an entry in G */
+            G_j[G_nnz++] = col - start;
+            if (G_nnz >= G_capacity)
             {
-               /* this is an entry in G */
-               G_j[G_nnz++] = col - start;
-               if (G_nnz >= G_capacity)
-               {
-                  HYPRE_Int tmp = G_capacity;
-                  G_capacity = (HYPRE_Int)(G_capacity * EXPAND_FACT + 1);
-                  G_j = hypre_TReAlloc_v2(G_j, HYPRE_Int, tmp, HYPRE_Int, G_capacity, HYPRE_MEMORY_DEVICE);
-               }
+               HYPRE_Int tmp = G_capacity;
+               G_capacity = (HYPRE_Int) (G_capacity * EXPAND_FACT + 1);
+               G_j = hypre_TReAlloc_v2(G_j, HYPRE_Int, tmp, HYPRE_Int,
+                                       G_capacity, HYPRE_MEMORY_HOST);
             }
          }
       }
-      G_i[num_nodes] = G_nnz;
-      if (G_nnz == 0)
```

Comment:
I am not sure about this part of the code, which seems to be gone. Was it unnecessary?

Reply:
In the new code, when `G_nnz == 0`, I am still freeing the variables `G_j`, `rqperm` and `G`. But that happens later on. I reorganized the logic to have less lines of code

Reply:
OK. How about the else branch, like `A + A'` ? Is it gone as well?

Reply:
Yes, I reorganized the code so that the `else` block from `if (sym)` is taken into account as well

Reply:
Ok . I guess it’s fine as long as the regression is good. I couldn’t follow all the details.